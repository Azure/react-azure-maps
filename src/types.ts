import { ComponentClass, CSSProperties, ReactElement, StatelessComponent } from 'react'
import atlas, {
  AnimationOptions,
  CameraBoundsOptions,
  CameraOptions,
  DataSourceOptions,
  HeatMapLayerOptions,
  HtmlMarker,
  HtmlMarkerEvents,
  HtmlMarkerOptions,
  ImageLayerOptions,
  LineLayerOptions,
  Map,
  MapDataEvent,
  MapErrorEvent,
  MapEvent,
  MapMouseEvent,
  MapMouseWheelEvent,
  MapTouchEvent,
  Options,
  PolygonExtrusionLayerOptions,
  PolygonLayerOptions,
  PopupEvents,
  PopupOptions,
  ServiceOptions,
  Shape,
  StyleOptions,
  SymbolLayerOptions,
  TargetedEvent,
  TileLayerOptions,
  TrafficOptions,
  UserInteractionOptions,
  Control,
  BubbleLayerOptions,
  LayerOptions,
  VectorTileSourceOptions
} from 'azure-maps-control'
import { drawing, DrawingManagerOptions, DrawingToolbarOptions } from 'azure-maps-drawing-tools'

export type IAzureMapOptions = ServiceOptions &
  StyleOptions &
  UserInteractionOptions &
  (CameraOptions | CameraBoundsOptions)

export type IAzureMapChildren =
  | ReactElement<IAzureMapHtmlMarker>
  | ReactElement<IAzureMapPopup>
  | ReactElement<IAzureMapDataSourceProps>

export type IAzureMap = {
  children?: Array<IAzureMapChildren> | IAzureMapChildren
  providedMapId?: string
  containerClassName?: string
  styles?: CSSProperties
  LoaderComponent?: ComponentClass<any> | StatelessComponent<any>
  options?: IAzureMapOptions
  imageSprites?: IAzureMapImageSprite[]
  controls?: IAzureMapControls[]
  customControls?: IAzureCustomControls[]
  events?: IAzureMapEvent | any
  cameraOptions?: AzureSetCameraOptions
  trafficOptions?: TrafficOptions
  userInteraction?: UserInteractionOptions
  styleOptions?: StyleOptions
  serviceOptions?: ServiceOptions
}
export type IAzureCustomControls = {
  control: Control
  controlOptions?: ControlOptions
}

export type IAzureMapControls = {
  controlName: string
  controlOptions?: Options
  options?: ControlOptions | undefined
}

export type IAzureMapImageSprite = {
  id: string
  templateName?: string
  color?: string
  secondaryColor?: string
  scale?: number
  icon?: string | HTMLImageElement | ImageData
}

export type IAzureMapContextState = {
  mapRef: Map | null
  isMapReady: boolean | false
  setMapRef(mapRef: Map): void
  removeMapRef(): void
  setMapReady(isMapReady: boolean): void
}

export type IAzureMapHtmlMarkerEvent = {
  eventName: keyof HtmlMarkerEvents
  callback: (e: TargetedEvent) => void
}

export type IAzureMapPopupEvent = {
  eventName: keyof PopupEvents
  callback: (e: TargetedEvent) => void
}

export type IAzureMapMouseEvents = {
  [T in keyof HtmlMarkerEvents]: (e: TargetedEvent) => void
}

export type IAzureMapHtmlMarker = {
  id?: string
  isPopupVisible?: boolean
  markerContent?: ReactElement
  options?: HtmlMarkerOptions
  events?: Array<IAzureMapHtmlMarkerEvent>
}

export type IAzureMapPopup = {
  isVisible?: boolean
  options?: PopupOptions
  events?: Array<IAzureMapPopupEvent>
  popupContent: ReactElement
}

export type IAzureMapDataSourceContextState = {
  dataSourceRef: atlas.source.DataSource | atlas.source.VectorTileSource | null
}

export type IAzureMapLayerContextState = {
  layerRef: atlas.layer.SymbolLayer | atlas.layer.ImageLayer | atlas.layer.TileLayer | null
}

export type IAzureDataSourceChildren =
  | IAzureMapFeature
  | ReactElement<IAzureMapFeature>
  | ReactElement<IAzureLayerStatefulProviderProps>

export type IAzureVectorTileSourceChildren = ReactElement<IAzureLayerStatefulProviderProps>

export type IAzureMapDataSourceEvent = {
  [property in IAzureMapDataSourceEventType]: (e: Shape[]) => void
}

export type IAzureMapVectorTileSourceEvent = {
  [property in IAzureMapSourceEventType]?: (e: atlas.source.VectorTileSource) => void
}

export type IAzureMapEvent = {
  [property in IAzureMapEventsType]: (
    e:
      | MapDataEvent
      | MapErrorEvent
      | MapTouchEvent
      | MapMouseEvent
      | string
      | MapMouseWheelEvent
      | MapEvent
      | atlas.layer.Layer
      | atlas.source.Source
  ) => void
}

export type IAzureDataSourceStatefulProviderProps = {
  id: string
  children?: | Array<IAzureDataSourceChildren | IAzureDataSourceChildren[] | null>
  | IAzureDataSourceChildren
  | null
  options?: DataSourceOptions
  events?: IAzureMapDataSourceEvent | any
  dataFromUrl?: string
  collection?:
    | atlas.data.FeatureCollection
    | atlas.data.Feature<atlas.data.Geometry, any>
    | atlas.data.Geometry
    | atlas.data.GeometryCollection
    | Shape
    | Array<atlas.data.Feature<atlas.data.Geometry, any> | atlas.data.Geometry | Shape>
  index?: number
}

export type IAzureVectorTileSourceStatefulProviderProps = {
  id: string,
  children?: | Array<IAzureVectorTileSourceChildren | IAzureVectorTileSourceChildren[] | null>
  | IAzureVectorTileSourceChildren
  | null
  options?: VectorTileSourceOptions,
  events?: IAzureMapVectorTileSourceEvent
  // NOTE: not sure yet why this is needed, haven't seen this used in AzureMapsDataSource, though IAzureGeoJSONDataSourceStatefulProviderProps has it
  index?: number
}

export type IAzureMapLayerEvent = {
  [property in IAzureMapLayerEventType]: (
    e: MapMouseEvent | MapTouchEvent | MapMouseWheelEvent
  ) => void
}

export type IAzureMapLifecycleEvent = {
  [property in IAzureMapLayerLifecycleEvents]: (e: atlas.layer.Layer) => void
}

export type IAzureLayerStatefulProviderProps = {
  id?: string
  options?:
    | (SymbolLayerOptions &
        HeatMapLayerOptions &
        ImageLayerOptions &
        LineLayerOptions &
        PolygonExtrusionLayerOptions &
        PolygonLayerOptions &
        TileLayerOptions &
        BubbleLayerOptions &
        LayerOptions)
    | Options
  type: IAzureMapLayerType
  events?: IAzureMapLayerEvent | any
  onCreateCustomLayer?: (dataSourceRef: DataSourceType, mapRef: MapType | null) => atlas.layer.Layer
  lifecycleEvents?: IAzureMapLifecycleEvent | any
}

export type IAzureDrawingLayerStatefulProviderProps = {
  // type of drawing layer to customize
  type: "lineLayer" | "polygonLayer" | "polygonOutlineLayer" | "pointLayer"
  options?:
    | (SymbolLayerOptions &
        LineLayerOptions &
        PolygonLayerOptions &
        LayerOptions)
    | Options
  events?: IAzureMapLayerEvent | any
  lifecycleEvents?: IAzureMapLifecycleEvent | any
}

export type IAzureMapLayerLifecycleEvents = 'layeradded' | 'layerremoved'

export type IAzureMapEventsType =
  | IAzureMapLayerEventType
  | IAzureMapLayerLifecycleEvents
  | IAzureMapDataSourceEventType
  | IAzureMapAddEventsType
  | IAzureMapSourceEventType
  // Adds a data event to the map.
  | 'data'
  | 'sourcedata'
  | 'styledata'
  // Adds an event to the map.
  | 'error'
  // Adds a style image missing event to the map.
  | 'styleimagemissing'

export type IAzureMapAddEventsType =
  | 'boxzoomstart'
  | 'boxzoomend'
  | 'dragstart'
  | 'drag'
  | 'dragend'
  | 'idle'
  | 'load'
  | 'movestart'
  | 'move'
  | 'moveend'
  | 'pitchstart'
  | 'pitch'
  | 'pitchend'
  | 'ready'
  | 'render'
  | 'resize'
  | 'rotatestart'
  | 'rotate'
  | 'rotateend'
  | 'tokenacquired'
  | 'zoomstart'
  | 'zoom'
  | 'zoomend'

export type IAzureMapDataSourceEventType = 'dataadded' | 'dataremoved'

export type IAzureMapSourceEventType = 'sourceadded' | 'sourceremoved'

export type IAzureMapLayerEventType =
  // Mouse events
  | 'mousedown'
  | 'mouseup'
  | 'mouseover'
  | 'mousemove'
  | 'click'
  | 'dblclick'
  | 'mouseout'
  | 'mouseenter'
  | 'mouseleave'
  | 'contextmenu'
  // Wheel events
  | 'wheel'
  // Touch events
  | 'touchstart'
  | 'touchend'
  | 'touchmove'
  | 'touchcancel'

export type IAzureMapLayerType =
  | 'SymbolLayer'
  | 'HeatLayer'
  | 'ImageLayer'
  | 'LineLayer'
  | 'PolygonExtrusionLayer'
  | 'PolygonLayer'
  | 'TileLayer'
  | 'BubbleLayer'
  | 'HtmlMarkerLayer'
  | 'custom'

export type IAzureMapFeatureType =
  | 'Point'
  | 'MultiPoint'
  | 'LineString'
  | 'MultiLineString'
  | 'Polygon'
  | 'MultiPolygon'

export type IAzureMapFeature = {
  id?: string
  type: IAzureMapFeatureType
  coordinate?: atlas.data.Position
  coordinates?: Array<atlas.data.Position>
  multipleCoordinates?: Array<Array<atlas.data.Position>>
  multipleDimensionCoordinates?: Array<Array<Array<atlas.data.Position>>>
  bbox?: atlas.data.BoundingBox
  variant?: IAzureMapFeatureVariant
  properties?: Options
  // Shape functions:
  setCoords?:
    | atlas.data.Position
    | atlas.data.Position[]
    | atlas.data.Position[][]
    | atlas.data.Position[][][]
  setProperties?: Options
}

// DRAWING MODULE:
export type DrawingManagerType = drawing.DrawingManager
export type IAzureDrawingManagerDrawingEventType = 
  | 'drawingchanged'
  | 'drawingchanging'
  | 'drawingcomplete'
  | 'drawingstarted'

export type IAzureDrawingManagerModeEventType = 'drawingmodechanged'
export type IAzureDrawingManagerEventType = IAzureDrawingManagerDrawingEventType | IAzureDrawingManagerModeEventType

export type IAzureDrawingManagerEvent = {
  [property in IAzureDrawingManagerDrawingEventType]?: (e: atlas.Shape) => void
} & {
  [property in IAzureDrawingManagerModeEventType]?: (a: drawing.DrawingMode) => void
}

export interface IAzureMapDrawingManagerProps {
  drawingManagerRef: drawing.DrawingManager | null
}

export interface IAzureDrawingManagerStatefulProviderProps {
  options: Omit<DrawingManagerOptions, 'toolbar'> & { 'toolbar'?: DrawingToolbarOptions }
  events?: IAzureDrawingManagerEvent
  children?: 
    | Array<IAzureDataSourceChildren | null> 
    | Array<ReactElement<IAzureDrawingLayerStatefulProviderProps> | null>
    | IAzureDataSourceChildren
    | ReactElement<IAzureDrawingLayerStatefulProviderProps>
    | null
}

export type IAzureMapLayerProps = IAzureMapLayerContextState
export type IAzureMapMouseEventRef = HtmlMarker // && other possible iterfaces
export type IAzureMapsContextProps = IAzureMapContextState
export type IAzureMapDataSourceProps = IAzureMapDataSourceContextState
export type DataSourceType = atlas.source.DataSource | atlas.source.VectorTileSource
export type LayerType = atlas.layer.SymbolLayer | atlas.layer.ImageLayer | atlas.layer.TileLayer
export type MapType = atlas.Map
export type GeometryType = atlas.data.Geometry
export type FeatureType = atlas.data.Feature<atlas.data.Geometry, any>
export type ShapeType = atlas.Shape
export type IAzureMapFeatureVariant = 'shape' | 'feature'

// Azure types
export type AzureDataLineString = atlas.data.LineString
export type AzureDataPosition = atlas.data.Position
export type ControlOptions = atlas.ControlOptions
export type AzureSetCameraOptions = ((CameraOptions | CameraBoundsOptions) & AnimationOptions) | any
