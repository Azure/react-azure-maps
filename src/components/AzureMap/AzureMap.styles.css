.hidden-accessible-element {
  display: none;
}

.atlas-map {
  overflow: unset !important;
}
