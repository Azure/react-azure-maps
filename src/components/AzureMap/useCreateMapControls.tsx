import { IAzureCustomControls, IAzureMapControls, MapType } from '../../types'
import atlas, {
  CompassControlOptions,
  ControlOptions,
  PitchControlOptions,
  StyleControlOptions,
  TrafficControlOptions,
  ZoomControlOptions
} from 'azure-maps-control'



export const useCreateMapControls = (mapRef: MapType, controls: IAzureMapControls[]) => {
  controls.forEach((control: IAzureMapControls) => {
    const { controlName, options, controlOptions } = control
    mapRef.controls.add(
      createControl(controlName, controlOptions) as atlas.ControlBase,
      options as ControlOptions
    )
  })
}

export const createControl = (
  type: string,
  options?: ControlOptions
): atlas.ControlBase | undefined => {
  switch (type) {
    case 'CompassControl':
      return new atlas.control.CompassControl(options as CompassControlOptions)
    case 'PitchControl':
      return new atlas.control.PitchControl(options as PitchControlOptions)
    case 'StyleControl':
      return new atlas.control.StyleControl(options as StyleControlOptions)
    case 'ZoomControl':
      return new atlas.control.ZoomControl(options as ZoomControlOptions)
    case 'TrafficControl':
      return new atlas.control.TrafficControl(options as TrafficControlOptions)
    case 'TrafficLegendControl':
      return new atlas.control.TrafficLegendControl()
    default:
      console.warn('Check the type and passed props properties or try CustomControl')
  }
}

export const useCreateMapCustomControls = (
  mapRef: MapType,
  customControls: IAzureCustomControls[]
) => {
  customControls.forEach(({ control, controlOptions }: IAzureCustomControls) => {
    mapRef.controls.add(control, controlOptions)
  })
}

