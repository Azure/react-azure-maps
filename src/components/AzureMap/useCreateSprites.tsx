import { IAzureMapImageSprite, MapType } from '../../types'

export const createImageSprites = async (
  mapRef: MapType,
  spriteArray: IAzureMapImageSprite[]
) => {
  await Promise.all(
    spriteArray.map(
      async ({ id, templateName, color, secondaryColor, scale, icon }: IAzureMapImageSprite) => {
        if (icon) {
          // Add an icon image to the map's image sprite for use with symbols and patterns.
          await mapRef.imageSprite.add(id, icon)
        }
        // Creates and adds an image to the maps image sprite. Reference this in the Polygon or Symbol layer.
        return mapRef.imageSprite.createFromTemplate(
          id,
          templateName || 'marker',
          color,
          secondaryColor,
          scale
        )
      }
    )
  )
}
