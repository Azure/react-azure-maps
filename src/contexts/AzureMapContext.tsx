import React, { createContext, ReactElement, useState ,useContext} from 'react'
import { Map } from 'azure-maps-control'
import { IAzureMap, IAzureMapsContextProps } from '../types'

export const AzureMapsContext = createContext<IAzureMapsContextProps>({
  mapRef: null,
  isMapReady: false,
  setMapRef: (mapRef: Map) => {},
  removeMapRef: () => {},
  setMapReady: () => {}
})
const { Provider, Consumer: AzureMapsConsumer } = AzureMapsContext

type IAzureMapsStatefulProviderProps = {
  children?: ReactElement<IAzureMap>
}

const AzureMapsStatefulProvider = ({ children }: IAzureMapsStatefulProviderProps) => {
  const [mapRef, setMapRef] = useState<Map | null>(null)
  const [isMapReady, setIsMapReady] = useState(false)

  return (
    <Provider
      value={{
        mapRef,
        setMapRef,
        isMapReady,
        setMapReady: setIsMapReady,
        removeMapRef: () => setMapRef(null)
      }}
    >
      {children}
    </Provider>
  )
}

const useAzureMaps = () => useContext<IAzureMapsContextProps>(AzureMapsContext)

export { AzureMapsConsumer, AzureMapsStatefulProvider as AzureMapsProvider, useAzureMaps }
