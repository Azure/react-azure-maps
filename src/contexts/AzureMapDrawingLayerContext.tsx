import React, { createContext } from 'react'
import { useAzureMapDrawingLayer } from '../hooks/useAzureMapDrawingLayer'
import { IAzureDrawingLayerStatefulProviderProps, IAzureMapLayerProps } from '../types'

const AzureMapDrawingLayerContext = createContext<IAzureMapLayerProps>({
  layerRef: null
})
const { Provider, Consumer: AzureMapDrawingLayerConsumer } = AzureMapDrawingLayerContext

const AzureMapDrawingLayerStatefulProvider = ({
  options,
  type,
  events,
  lifecycleEvents
}: IAzureDrawingLayerStatefulProviderProps) => {
  const { layerRef } = useAzureMapDrawingLayer({
    options,
    type,
    events,
    lifecycleEvents
  })

  return (
    <Provider
      value={{
        layerRef
      }}
    ></Provider>
  )
}

export {
  AzureMapDrawingLayerContext,
  AzureMapDrawingLayerConsumer,
  AzureMapDrawingLayerStatefulProvider as AzureMapDrawingLayerProvider
}
