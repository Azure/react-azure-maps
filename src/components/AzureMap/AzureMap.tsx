import React, { memo, useContext, useEffect, useState } from 'react'
import atlas from 'azure-maps-control'
import { IAzureMap, IAzureMapsContextProps, MapType } from '../../types'
import { AzureMapsContext } from '../../contexts/AzureMapContext'
import { Guid } from 'guid-typescript'
import 'azure-maps-control/dist/atlas.min.css'
import 'mapbox-gl/src/css/mapbox-gl.css'
import { useCheckRef } from '../../hooks/useCheckRef'
import { createImageSprites } from './useCreateSprites'
import { createMapControls, createMapCustomControls } from './useCreateMapControls'

const AzureMap = memo(
  ({
    children, // @TODO We need to cover and type all possible childrens that we can pass to this component as child for. ex. Markers etc
    LoaderComponent = () => <div>Loading ...</div>,
    providedMapId,
    containerClassName,
    styles,
    options = {},
    imageSprites,
    controls,
    customControls,
    events,
    cameraOptions,
    trafficOptions,
    userInteraction,
    styleOptions,
    serviceOptions
  }: IAzureMap) => {
    const { setMapRef, removeMapRef, mapRef, setMapReady, isMapReady } = useContext<
      IAzureMapsContextProps
    >(AzureMapsContext)
    const [mapId] = useState(providedMapId || Guid.create().toString())
    useEffect(() => {
      if (mapRef) {
        mapRef.setTraffic(trafficOptions)
      }
    }, [trafficOptions])

    useEffect(() => {
      if (mapRef) {
        mapRef.setUserInteraction(userInteraction)
      }
    }, [userInteraction])

    useEffect(() => {
      if (mapRef) {
        mapRef.setCamera(cameraOptions)
      }
    }, [cameraOptions])

    useEffect(() => {
      if (mapRef) {
        mapRef.setStyle(styleOptions)
      }
    }, [styleOptions])

    useEffect(() => {
      if (mapRef && serviceOptions) {
        mapRef.setServiceOptions(serviceOptions)
      }
    }, [serviceOptions])

    useCheckRef<MapType, MapType>(mapRef, mapRef, mref => {
      mref.events.add('ready', () => {
        if (imageSprites) {
          createImageSprites(mref, imageSprites)
        }
        if (controls) {
          createMapControls(mref, controls)
        }
        if (customControls) {
          createMapCustomControls(mref, customControls)
        }
        if (trafficOptions) {
          mref.setTraffic(trafficOptions)
        }
        if (userInteraction) {
          mref.setUserInteraction(userInteraction)
        }
        if (cameraOptions) {
          mref.setCamera(cameraOptions)
        }
        if (styleOptions) {
          mref.setStyle(styleOptions)
        }
        if (serviceOptions) {
          mref.setServiceOptions(serviceOptions)
        }
        setMapReady(true)
      })
      for (const eventType in events) {
        mref.events.add(eventType as any, events[eventType])
      }
    })

    useEffect(() => {
      setMapRef(new atlas.Map(mapId, options))
      return () => {
        removeMapRef()
      }
    }, [])

    return (
      <>
        {!isMapReady && LoaderComponent && <LoaderComponent />}
        <div className={containerClassName} id={mapId} style={{ ...styles, height: '100%' }}>
          {isMapReady && children}
        </div>
      </>
    )
  }
)

export default AzureMap
