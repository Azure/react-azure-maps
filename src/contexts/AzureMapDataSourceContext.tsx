import React, { createContext, useContext, useEffect, useState } from 'react'
import {
  DataSourceType,
  IAzureDataSourceStatefulProviderProps,
  IAzureMapDataSourceProps,
  IAzureMapsContextProps,
  MapType
} from '../types'
import atlas from 'azure-maps-control'
import { AzureMapsContext } from './AzureMapContext'
import { useCheckRef } from '../hooks/useCheckRef'

const AzureMapDataSourceContext = createContext<IAzureMapDataSourceProps>({
  dataSourceRef: null
})
const { Provider, Consumer: AzureMapDataSourceConsumer } = AzureMapDataSourceContext

/**
 * @param id
 * @param children
 * @param options
 * @param events
 * @param dataFromUrl Async Data from URL
 * @param collection Use for contain Feature Collection !All Feature child will be ignored when collection value will change
 */
const AzureMapDataSourceStatefulProvider = ({
  id,
  children,
  options,
  events,
  dataFromUrl,
  collection
}: IAzureDataSourceStatefulProviderProps) => {
  const [dataSourceRef] = useState<atlas.source.DataSource>(new atlas.source.DataSource(id, options))
  const { mapRef } = useContext<IAzureMapsContextProps>(AzureMapsContext)
  useCheckRef<MapType, DataSourceType>(mapRef, dataSourceRef, (mref, dref) => {
    for (const eventType in events || {}) {
      mref.events.add(eventType as any, dref, events[eventType])
    }
    mref.sources.add(dref)
    if (dref instanceof atlas.source.DataSource) {
      if (dataFromUrl) {
        dref.importDataFromUrl(dataFromUrl)
      }
      if (collection) {
        dref.add(collection)
      }
    }
  })

  useEffect(() => {
    if (dataSourceRef && collection) {
      // Cleared old values prevent duplication
      dataSourceRef.clear()
      dataSourceRef.add(collection)
    }
  }, [collection])

  useEffect(() => {
    if (dataSourceRef && options) {
      dataSourceRef.setOptions(options)
    }
  }, [options])

  return (
    <Provider
      value={{
        dataSourceRef
      }}
    >
      {mapRef && children}
    </Provider>
  )
}

export {
  AzureMapDataSourceContext,
  AzureMapDataSourceConsumer,
  AzureMapDataSourceStatefulProvider as AzureMapDataSourceProvider,
  Provider as AzureMapDataSourceRawProvider
}
