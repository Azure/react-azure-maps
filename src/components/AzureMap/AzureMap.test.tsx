import React from 'react'
import { render, act } from '@testing-library/react'
import { AzureMapsContext } from '../../contexts/AzureMapContext'
import AzureMap from './AzureMap'
import { Map } from 'azure-maps-control'
import { IAzureMap, IAzureMapsContextProps } from '../../types'
import { createImageSprites } from './useCreateSprites'
import { createMapCustomControls, createMapControls } from './useCreateMapControls'

const LoaderComponent = () => <div>Loader</div>

jest.mock('./useCreateMapControls', () => {
  return {
    createMapCustomControls: jest.fn(),
    createMapControls: jest.fn()
  }
})

jest.mock('./useCreateSprites')
jest.mock('guid-typescript', () => {
  return {
    Guid: {
      create: jest.fn(() => 'fake_generated_id')
    }
  }
})

const mapContextProps = {
  mapRef: null,
  isMapReady: false,
  setMapReady: jest.fn(),
  removeMapRef: jest.fn(),
  setMapRef: jest.fn()
}

const wrapWithAzureMapContext = (mapContextProps: IAzureMapsContextProps, mapProps: IAzureMap) => {
  return (
    <AzureMapsContext.Provider
      value={{
        ...mapContextProps
      }}
    >
      <AzureMap {...mapProps} />
    </AzureMapsContext.Provider>
  )
}

describe('AzureMap Component', () => {
  beforeEach(() => {
    mapContextProps.removeMapRef.mockClear()
    mapContextProps.setMapReady.mockClear()
    mapContextProps.setMapRef.mockClear()
  })

  it('should setMapRef on mount', () => {
    act(() => {
      render(wrapWithAzureMapContext(mapContextProps, {}))
    })
    expect(mapContextProps.setMapRef).toHaveBeenCalled()
  })

  it('should change trafficOptions call setTraffic from mapRef', () => {
    const mapRef = new Map('fake', {})
    act(() => {
      const { rerender } = render(wrapWithAzureMapContext({ ...mapContextProps, mapRef }, {}))
      rerender(
        wrapWithAzureMapContext(
          { ...mapContextProps, mapRef },
          { trafficOptions: { some: 'some2' } }
        )
      )
    })
    expect(mapRef.setTraffic).toHaveBeenCalledWith({ some: 'some2' })
  })

  it('should change userInteraction call setUserInteraction from mapRef', () => {
    const mapRef = new Map('fake', {})
    act(() => {
      const { rerender } = render(wrapWithAzureMapContext({ ...mapContextProps, mapRef }, {}))
      rerender(
        wrapWithAzureMapContext(
          { ...mapContextProps, mapRef },
          { userInteraction: { some: 'some2' } }
        )
      )
    })
    expect(mapRef.setUserInteraction).toHaveBeenCalledWith({ some: 'some2' })
  })

  it('should change cameraOptions call setCamera from mapRef', () => {
    const mapRef = new Map('fake', {})
    act(() => {
      const { rerender } = render(wrapWithAzureMapContext({ ...mapContextProps, mapRef }, {}))
      rerender(
        wrapWithAzureMapContext(
          { ...mapContextProps, mapRef },
          { cameraOptions: { some: 'some2' } }
        )
      )
    })
    expect(mapRef.setCamera).toHaveBeenCalledWith({ some: 'some2' })
  })

  it('should call removeMapRef on unmount of component', () => {
    const mapRef = new Map('fake', {})
    const { unmount } = render(wrapWithAzureMapContext({ ...mapContextProps, mapRef }, {}))
    unmount()
    expect(mapContextProps.removeMapRef).toHaveBeenCalled()
  })

  it('should call createImageSprites if imageSprites is not falsy', () => {
    const mapRef = new Map('fake', {})
    render(
      wrapWithAzureMapContext(
        { ...mapContextProps, mapRef },
        { imageSprites: [{ id: 'some_fake_id' }] }
      )
    )
    expect(createImageSprites).toHaveBeenCalled()
  })

  it('should call createMapControls if controls is not falsy', () => {
    const mapRef = new Map('fake', {})
    const fakeControls = [{ controlName: 'fake_control_name' }]
    render(wrapWithAzureMapContext({ ...mapContextProps, mapRef }, { controls: fakeControls }))
    expect(createMapControls).toHaveBeenCalledWith(expect.any(Object), fakeControls)
  })

  it('should call createMapCustomControls if customControls is not falsy', () => {
    const mapRef = new Map('fake', {})
    const customControls = [
      {
        control: { onAdd: jest.fn(), onRemove: jest.fn() },
        controlOptions: {}
      }
    ]
    render(
      wrapWithAzureMapContext(
        { ...mapContextProps, mapRef },
        {
          customControls
        }
      )
    )
    expect(createMapCustomControls).toHaveBeenCalledWith(expect.any(Object), customControls)
  })

  it('should setTraffic on initial props', () => {
    const mapRef = new Map('fake', {})
    render(
      wrapWithAzureMapContext({ ...mapContextProps, mapRef }, { trafficOptions: { some: 'some2' } })
    )
    expect(mapRef.setTraffic).toHaveBeenCalledWith({ some: 'some2' })
  })

  it('should userInteraction on initial props', () => {
    const mapRef = new Map('fake', {})
    render(
      wrapWithAzureMapContext(
        { ...mapContextProps, mapRef },
        { userInteraction: { some: 'some2' } }
      )
    )
    expect(mapRef.setUserInteraction).toHaveBeenCalledWith({ some: 'some2' })
  })

  it('should cameraOptions on initial props', () => {
    const mapRef = new Map('fake', {})
    render(
      wrapWithAzureMapContext({ ...mapContextProps, mapRef }, { cameraOptions: { some: 'some2' } })
    )
    expect(mapRef.setCamera).toHaveBeenCalledWith({ some: 'some2' })
  })

  it('should setStyle on initial props', () => {
    const mapRef = new Map('fake', {})
    render(
      wrapWithAzureMapContext({ ...mapContextProps, mapRef }, { styleOptions: { some: 'some2' } })
    )
    expect(mapRef.setStyle).toHaveBeenCalledWith({ some: 'some2' })
  })

  it('should setServiceOptions on initial props', () => {
    const mapRef = new Map('fake', {})
    render(
      wrapWithAzureMapContext({ ...mapContextProps, mapRef }, { serviceOptions: { some: 'some2' } })
    )
    expect(mapRef.setServiceOptions).toHaveBeenCalledWith({ some: 'some2' })
  })

  it('should call setMapready on mount of component', () => {
    const mapRef = new Map('fake', {})
    render(wrapWithAzureMapContext({ ...mapContextProps, mapRef }, {}))
    expect(mapContextProps.setMapReady).toHaveBeenCalledWith(true)
  })

  it('should add props events to mapRef', () => {
    const mapRef = new Map('fake', { options: {} })
    const dataCallback = () => {
      console.log('some fake text')
    }
    render(
      wrapWithAzureMapContext(
        { ...mapContextProps, mapRef },
        {
          events: {
            data: dataCallback
          }
        }
      )
    )
    expect(mapRef.events.add).toHaveBeenCalledWith('ready', expect.any(Function))
    expect(mapRef.events.add).toHaveBeenCalledWith('data', dataCallback)
  })

  it('should render LoaderComponent if isMapReady is false and LoaderComponent exists', async () => {
    const mapRef = new Map('fake', { options: {} })
    const { findByText } = render(
      wrapWithAzureMapContext(
        { ...mapContextProps, mapRef },
        {
          LoaderComponent
        }
      )
    )
    const loaderElement = await findByText('Loader')
    expect(loaderElement).toMatchSnapshot()
  })

  it('should create map with div and automatically generated id when if isMapReady is true and LoaderComponent exists', async () => {
    const mapRef = new Map('fake', { options: {} })
    const { container } = render(
      wrapWithAzureMapContext({ ...mapContextProps, mapRef, isMapReady: true }, {})
    )
    expect(container).toMatchSnapshot()
  })

  it('should render map with div and provvided id when if isMapReady is true and LoaderComponent exists', async () => {
    const mapRef = new Map('fake', { options: {} })
    const { container } = render(
      wrapWithAzureMapContext(
        { ...mapContextProps, mapRef, isMapReady: true },
        {
          LoaderComponent,
          providedMapId: 'some_fake_map_id'
        }
      )
    )
    expect(container).toMatchSnapshot()
  })

  afterAll(() => {
    jest.unmock('./useCreateSprites')
    jest.unmock('./useCreateMapControls')
    jest.unmock('guid-typescript')
  })
})
