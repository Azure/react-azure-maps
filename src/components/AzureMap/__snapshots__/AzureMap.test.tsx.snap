// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AzureMap Component should create map with div and automatically generated id when if isMapReady is true and LoaderComponent exists 1`] = `
<div>
  <div
    id="fake_generated_id"
    style="height: 100%;"
  />
</div>
`;

exports[`AzureMap Component should render LoaderComponent if isMapReady is false and LoaderComponent exists 1`] = `
<div>
  Loader
</div>
`;

exports[`AzureMap Component should render map with div and provvided id when if isMapReady is true and LoaderComponent exists 1`] = `
<div>
  <div
    id="some_fake_map_id"
    style="height: 100%;"
  />
</div>
`;
